File: Backend/routes/books.py
from fastapi import (
    APIRouter,
    UploadFile,
    File,
    Form,
    Depends,
    Query,
    BackgroundTasks,
)
from utils.auth_helper import get_user
from utils.admin_helper import get_admin_user
from fastapi.encoders import jsonable_encoder

from services.bookService import (
    get_books,
    get_random_books,
    get_book_by_id,
    create_book,
    update_book,
    delete_book,
)
from schemas.books import BookCreate, BookUpdate
from socketio_app import sio

# ❗❗❗ אין prefix כאן
router = APIRouter(tags=["Book"])


# =========================
# CREATE BOOK (ADMIN)
# =========================
@router.post("/")
async def add_book(
    admin=Depends(get_admin_user),
    title: str = Form(...),
    summary: str = Form(...),
    author: str = Form(...),
    quantity: int = Form(...),
    pages: int = Form(...),
    categoryid: int = Form(...),
    agesid: int = Form(...),
    image: UploadFile = File(...),
):
    data = BookCreate(
        title=title,
        summary=summary,
        author=author,
        quantity=quantity,
        pages=pages,
        categoryid=categoryid,
        agesid=agesid,
    )

    new_book = create_book(data, image)

    await sio.emit(
        "books_changed",
        jsonable_encoder({
            "reason": "created",
            "book": new_book,   # 👈 send full book
            "userId": admin["id"],
        }),
    )

    return new_book



# =========================
# UPDATE BOOK (ADMIN)
# =========================
@router.put("/{book_id}")
async def edit_book(
    book_id: int,
    admin=Depends(get_admin_user),
    title: str | None = Form(None),
    summary: str | None = Form(None),
    author: str | None = Form(None),
    quantity: int | None = Form(None),
    pages: int | None = Form(None),
    categoryid: int | None = Form(None),
    agesid: int | None = Form(None),
    image: UploadFile | None = File(None),
):
    data = BookUpdate(
        title=title,
        summary=summary,
        author=author,
        quantity=quantity,
        pages=pages,
        categoryid=categoryid,
        agesid=agesid,
    )

    updated = update_book(book_id, data, image)

    await sio.emit(
        "books_changed",
        jsonable_encoder({
            "reason": "updated",
            "book": updated,
            "userId": admin["id"]
        }),
    )

    return updated



# =========================
# DELETE BOOK (ADMIN)
# =========================
@router.delete("/{book_id}")
async def remove_book(
    book_id: int,
    admin=Depends(get_admin_user),
):
    result = delete_book(book_id)

    await sio.emit(
        "books_changed",
        {
            "reason": "deleted",
            "bookId": book_id,
            "userId": admin["id"],
        },
    )

    return result



# =========================
# LIST BOOKS
# =========================
@router.get("/")
def list_books(
    page: int = Query(1, ge=1),
    limit: int = Query(10, ge=1, le=50),
    category_id: int | None = Query(None),
    age_group_id: int | None = Query(None),
    search: str | None = Query(None),
    user=Depends(get_user),
):
    return get_books(page, limit, category_id, age_group_id, search)


# =========================
# RANDOM BOOKS
# =========================
@router.get("/random/limit")
def random_books(
    limit: int = Query(10, ge=1, le=50),
    user=Depends(get_user),
):
    return get_random_books(limit)


# =========================
# SINGLE BOOK
# =========================
@router.get("/{book_id}")
def single_book(book_id: int, user=Depends(get_user)):
    return get_book_by_id(book_id)

File: Frontend/src/components/services/books.js
// services/books.js
import axios from "axios";
const APIKEY = import.meta.env.VITE_API_KEY;


const BASE_URL = "http://localhost:8000/api/book";

class Books {
  static async getBooks(
    page = 1,
    limit = 8,
    categoryId = null,
    ageGroupId = null,
    search = ""
  ) {
    const res = await axios.get(BASE_URL, {
      withCredentials: true,
      params: {
        page,
        limit,
        category_id: categoryId,
        age_group_id: ageGroupId,
        search: search || undefined,
      },
      headers: { apiKey: APIKEY },
    });
    return res.data;
  }

  static async getBookById(id) {
    const res = await axios.get(`${BASE_URL}/${id}`, {
      withCredentials: true,
      headers: { apiKey: APIKEY },
    });
    return res.data;
  }

  static async getRandomBooks(limit = 10) {
    const res = await axios.get(`${BASE_URL}/random/limit`, {
      params: { limit },
      withCredentials: true,
      headers: { apiKey:APIKEY },
    });
    return res.data;
  }

  static async addBook(data) {
    const formData = new FormData();

    formData.append("title", data.title);
    formData.append("summary", data.summary);
    formData.append("author", data.author);
    formData.append("quantity", data.quantity);
    formData.append("pages", data.pages);
    formData.append("categoryid", data.categoryid);
    formData.append("agesid", data.agesid);

    if (data.imageFile) {
      formData.append("image", data.imageFile);
    }

    const res = await axios.post(BASE_URL, formData, {
      withCredentials: true,
      headers: { apiKey: APIKEY },
    });

    return res.data;
  }

  static async updateBook(id, data) {
    const formData = new FormData();

    if (data.title !== undefined) formData.append("title", data.title);
    if (data.summary !== undefined) formData.append("summary", data.summary);
    if (data.author !== undefined) formData.append("author", data.author);
    if (data.quantity !== undefined) formData.append("quantity", data.quantity);
    if (data.pages !== undefined) formData.append("pages", data.pages);
    if (data.categoryid !== undefined)
      formData.append("categoryid", data.categoryid);
    if (data.agesid !== undefined) formData.append("agesid", data.agesid);

    if (data.imageFile) {
      formData.append("image", data.imageFile);
    }

    const res = await axios.put(`${BASE_URL}/${id}`, formData, {
      withCredentials: true,
      headers: { apiKey:APIKEY },
    });

    return res.data;
  }

  static async deleteBook(id) {
    await axios.delete(`${BASE_URL}/${id}`, {
      withCredentials: true,
      headers: { apiKey:APIKEY },
    });
  }
}

export default Books;
